### ISSUE TITLE
Improve sort performance for matrices and volumes

### ISSUE THREAD

**Issue by pavanky** [arrayfire/arrayfire#1358]
Right now we are performing `N` sorts of size `M`  for an `M x N` matrix.

A faster solution might be to do the following internally:

``` cpp
af::array sort2(af::array val)
{
    af::array key = af::range(val.dims(), 1);
    af::array resval, reskey;
    af::sort(resval, reskey, af::flat(val), af::flat(key));

    af::array finalval, finalkey;
    af::sort(finalkey, finalval, reskey, resval);
    return af::moddims(finalval, val.dims());
}
```

This requires about `3XMxN` temporary storage. But that might be OK upto certain sizes. We might have to do a hybrid algorithm that does this both the for loop and this for larger sizes.

**Comment by pavanky** [comment]
This can probably be generalized for sorting along other dimensions using the following code:

```
af::array sort2(af::array val, int d = 0)
{   
    af::dim4 seq_dims = val.dims();
    af::dim4 tile_dims = af::dim4(1,1,1,1);

    tile_dims[d] = seq_dims[d];
    seq_dims[d] = 1;

    af::array key = af::iota(seq_dims, tile_dims);
    af::array resval, reskey;
    af::sort(resval, reskey, af::flat(val), af::flat(key));

    af::array finalval, finalkey;
    af::sort(finalkey, finalval, reskey, resval);
    return af::moddims(finalval, val.dims());
}
```

**Comment by pavanky** [comment]
Reference benchmark using the following code a GTX 950:
```cpp
int main(void)
{   
    af::array a = randu(1024, 4096);
    af::array b = sort(a);
    af::array c = sort2(a);

    printf("%f\n", max<float>(abs(b - c)));

    int niter = 5;

    timer::start();
    for (int i = 0; i < niter; i++) {
        b = sort(a);
    }
    af::sync();
    printf("Time for af::sort(): %lf\n", timer::stop() / niter);


    timer::start();
    for (int i = 0; i < niter; i++) {
        c = sort2(a);
    }
    af::sync();
    printf("Time for af::sort2(): %lf\n", timer::stop() / niter);

    return 0;
}
```

```
$ ./build/examples/helloworld/helloworld_cuda
0.000000
Time for af::sort(): 1.203843
Time for af::sort2(): 0.017735
```

**Comment by pavanky** [comment]
Fixed in hotfix-3.3.2
